File: weird_captcha_gym/shared_runtime/app/mechanics/teach_the_stencil.js
(() => {
  "use strict";
  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};

  const PALETTE = ["#ead8ae", "#6fbf78", "#b48555", "#ee8eb0", "#f1c963", "#9b8ba9"];
  const MIN_FULL_PATH_LENGTH = 4.0;
  const BRUSH_OFFSETS = {
    1: [[0, 0]],
    2: [[0, 0], [1, 0]],
    3: [[0, 0], [1, 0], [-1, 0], [0, 1], [0, -1]],
    4: [[0, 0], [1, 0], [-1, 0], [0, 1], [0, -1], [1, 1], [-1, -1]],
    5: [[0, 0], [1, 0], [-1, 0], [0, 1], [0, -1], [1, 1], [-1, -1], [1, -1], [-1, 1], [2, 0], [-2, 0], [0, 2], [0, -2]],
  };
  const model = {
    state: null,
    helpers: null,
    canvas: null,
    ctx: null,
    plot: {x: 26, y: 20, width: 848, height: 440},
    selectedClass: null,
    tool: "paint",
    brushSize: 2,
    samples: [],
    predictions: null,
    events: [],
    pointer: null,
    drawing: false,
    submitting: false,
    updateCount: 0,
    verdict: "idle",
  };

  function interaction() {
    return String(model.state?.control_condition?.interaction || "simplified");
  }

  function sourceFor(type) {
    if (type === "paint") return interaction() === "full" ? "brush_drag" : "proxy_stamp";
    return interaction() === "full" ? "eraser_drag" : "proxy_erase";
  }

  function pointFromEvent(event) {
    if (!model.canvas || !model.state?.plate) return null;
    const rect = model.canvas.getBoundingClientRect();
    const sx = model.canvas.width / rect.width;
    const sy = model.canvas.height / rect.height;
    const px = (event.clientX - rect.left) * sx;
    const py = (event.clientY - rect.top) * sy;
    const x = (px - model.plot.x) * Number(model.state.plate.width) / model.plot.width;
    const y = (py - model.plot.y) * Number(model.state.plate.height) / model.plot.height;
    if (x < 0 || y < 0 || x >= Number(model.state.plate.width) || y >= Number(model.state.plate.height)) return null;
    return {x: Math.max(0, Math.min(Number(model.state.plate.width) - 0.01, Number(x.toFixed(2)))), y: Math.max(0, Math.min(Number(model.state.plate.height) - 0.01, Number(y.toFixed(2))))};
  }

  function addEvent(type, details) {
    model.events.push(Object.assign({seq: model.events.length + 1, type}, details || {}));
  }

  function featureAt(point) {
    const width = Number(model.state.plate.width);
    const index = Math.min(model.state.plate.pixels.length - 1, Math.max(0, Math.floor(point.y) * width + Math.floor(point.x)));
    return model.state.plate.pixels[index];
  }

  function brushSizeValue(value) {
    const size = Math.round(Number(value));
    return Number.isInteger(size) && BRUSH_OFFSETS[size] ? size : 1;
  }

  function brushPoints(point, size) {
    const width = Number(model.state.plate.width);
    const height = Number(model.state.plate.height);
    const baseX = Math.floor(Number(point.x));
    const baseY = Math.floor(Number(point.y));
    const result = [];
    for (const [offsetX, offsetY] of BRUSH_OFFSETS[size]) {
      const x = baseX + offsetX;
      const y = baseY + offsetY;
      if (x < 0 || y < 0 || x >= width || y >= height) continue;
      if (!result.some((item) => item.x === x && item.y === y)) result.push({x, y});
    }
    return result;
  }

  function eraserRadius(size) {
    return 0.8 + 0.55 * Number(size);
  }

  function pathLength(points) {
    let length = 0;
    for (let index = 1; index < points.length; index += 1) {
      length += Math.hypot(points[index].x - points[index - 1].x, points[index].y - points[index - 1].y);
    }
    return length;
  }

  function applyStroke(type, points, classId, brushSize = 1) {
    const size = brushSizeValue(brushSize);
    if (type === "paint") {
      for (const point of points) {
        for (const samplePoint of brushPoints(point, size)) {
          model.samples.push({class_id: classId, x: samplePoint.x, y: samplePoint.y, feature: featureAt(samplePoint)});
        }
      }
      return;
    }
    const radius = eraserRadius(size);
    model.samples = model.samples.filter((sample) => !points.some((point) => Math.hypot(Number(sample.x) - point.x, Number(sample.y) - point.y) <= radius));
  }

  function distance(feature, centroid) {
    let value = 0;
    for (let i = 0; i < 3; i += 1) value += (Number(feature[i]) / 255 - Number(centroid[i]) / 255) ** 2;
    value += 0.65 * (Number(feature[3]) - Number(centroid[3])) ** 2;
    return value;
  }

  function predict() {
    const classCount = model.state.classes.length;
    const centroids = [];
    for (let classId = 0; classId < classCount; classId += 1) {
      const points = model.samples.filter((sample) => Number(sample.class_id) === classId);
      if (!points.length) return null;
      centroids[classId] = [0, 0, 0, 0];
      for (const point of points) for (let channel = 0; channel < 4; channel += 1) centroids[classId][channel] += Number(point.feature[channel]);
      centroids[classId] = centroids[classId].map((value) => value / points.length);
    }
    return model.state.plate.pixels.map((feature) => {
      let best = 0;
      let bestDistance = Infinity;
      for (let classId = 0; classId < classCount; classId += 1) {
        const current = distance(feature, centroids[classId]);
        if (current < bestDistance) { best = classId; bestDistance = current; }
      }
      return best;
    });
  }

  function classColor(classId) {
    return model.state.classes[classId]?.swatch || PALETTE[classId % PALETTE.length];
  }

  function draw() {
    if (!model.ctx || !model.state) return;
    const ctx = model.ctx;
    const width = Number(model.state.plate.width);
    const height = Number(model.state.plate.height);
    ctx.clearRect(0, 0, model.canvas.width, model.canvas.height);
    const gradient = ctx.createLinearGradient(0, 0, model.canvas.width, model.canvas.height);
    gradient.addColorStop(0, "#192d31");
    gradient.addColorStop(1, "#0b181d");
    ctx.fillStyle = gradient;
    ctx.fillRect(0, 0, model.canvas.width, model.canvas.height);
    ctx.save();
    ctx.shadowColor = "rgba(0,0,0,.45)";
    ctx.shadowBlur = 24;
    ctx.fillStyle = "#d8cbb1";
    ctx.fillRect(model.plot.x - 8, model.plot.y - 8, model.plot.width + 16, model.plot.height + 16);
    ctx.restore();
    const cellWidth = model.plot.width / width;
    const cellHeight = model.plot.height / height;
    for (let y = 0; y < height; y += 1) {
      for (let x = 0; x < width; x += 1) {
        const feature = model.state.plate.pixels[y * width + x];
        ctx.fillStyle = `rgb(${Math.round(feature[0])},${Math.round(feature[1])},${Math.round(feature[2])})`;
        ctx.fillRect(model.plot.x + x * cellWidth, model.plot.y + y * cellHeight, cellWidth + 0.5, cellHeight + 0.5);
      }
    }
    if (Array.isArray(model.predictions)) {
      ctx.globalAlpha = 0.43;
      for (let y = 0; y < height; y += 1) {
        for (let x = 0; x < width; x += 1) {
          ctx.fillStyle = classColor(model.predictions[y * width + x]);
          ctx.fillRect(model.plot.x + x * cellWidth, model.plot.y + y * cellHeight, cellWidth + 0.5, cellHeight + 0.5);
        }
      }
      ctx.globalAlpha = 1;
    }
    for (const sample of model.samples) {
      const sx = model.plot.x + Number(sample.x) * cellWidth;
      const sy = model.plot.y + Number(sample.y) * cellHeight;
      ctx.beginPath();
      ctx.arc(sx, sy, Math.max(2.5, Number(model.brushSize) * 1.6), 0, Math.PI * 2);
      ctx.fillStyle = classColor(Number(sample.class_id));
      ctx.globalAlpha = 0.92;
      ctx.fill();
      ctx.strokeStyle = "rgba(15,29,29,.85)";
      ctx.lineWidth = 1;
      ctx.stroke();
      ctx.globalAlpha = 1;
    }
    ctx.strokeStyle = "rgba(255,245,213,.9)";
    ctx.lineWidth = 2;
    ctx.strokeRect(model.plot.x, model.plot.y, model.plot.width, model.plot.height);
  }

  function status(message, state = "idle") {
    model.helpers?.setReadout(message, state);
    const node = document.querySelector(".teach-stencil-verdict");
    if (node) { node.dataset.status = state; node.textContent = message; }
  }

  function chooseClass(classId) {
    model.selectedClass = Number(classId);
    document.querySelectorAll("[data-stencil-class]").forEach((button) => button.classList.toggle("is-selected", Number(button.dataset.stencilClass) === model.selectedClass));
    addEvent("select_class", {class_id: model.selectedClass, input_source: "palette_button"});
    status("ACTIVE", "active");
  }

  function setTool(tool) {
    model.tool = tool;
    document.querySelectorAll("[data-stencil-tool]").forEach((button) => button.classList.toggle("is-selected", button.dataset.stencilTool === tool));
    const canvas = document.querySelector("#teach-stencil-canvas");
    if (canvas) canvas.dataset.tool = tool;
  }

  function finishStroke(points) {
    if (!points.length) return;
    if (model.tool === "paint" && model.selectedClass == null) {
      status("FAIL", "error");
      return;
    }
    const type = model.tool === "paint" ? "paint" : "erase";
    if (interaction() === "full" && (points.length < 2 || pathLength(points) < MIN_FULL_PATH_LENGTH)) {
      status("FAIL", "error");
      return;
    }
    const size = brushSizeValue(model.brushSize);
    const details = {points, brush_size: size, input_source: sourceFor(type)};
    if (type === "paint") details.class_id = model.selectedClass;
    addEvent(type, details);
    applyStroke(type, points, model.selectedClass, size);
    draw();
    status(type === "paint" ? "MARKED" : "ERASED", "active");
  }

  function installPointerHandlers() {
    model.canvas.addEventListener("pointerdown", (event) => {
      const point = pointFromEvent(event);
      if (!point) return;
      model.pointer = point;
      model.drawing = true;
      model.canvas.setPointerCapture(event.pointerId);
      if (interaction() === "simplified") finishStroke([point]);
      else finishStroke([]);
      if (interaction() === "full") model.pointer = {points: [point], last: point, pointerId: event.pointerId};
    });
    model.canvas.addEventListener("pointermove", (event) => {
      if (!model.drawing || interaction() !== "full" || !model.pointer?.points) return;
      const point = pointFromEvent(event);
      if (!point) return;
      if (Math.hypot(point.x - model.pointer.last.x, point.y - model.pointer.last.y) >= 0.55) {
        model.pointer.points.push(point);
        model.pointer.last = point;
        draw();
      }
    });
    const finish = (event) => {
      if (!model.drawing || interaction() !== "full" || !model.pointer?.points) { model.drawing = false; return; }
      model.drawing = false;
      const points = model.pointer.points;
      model.pointer = null;
      finishStroke(points);
    };
    model.canvas.addEventListener("pointerup", finish);
    model.canvas.addEventListener("pointercancel", finish);
  }

  function recomputeFromEvents() {
    model.samples = [];
    model.predictions = null;
    model.updateCount = 0;
    for (const event of model.events) {
      if (event.type === "paint") applyStroke("paint", event.points || [], event.class_id, event.brush_size);
      else if (event.type === "erase") applyStroke("erase", event.points || [], null, event.brush_size);
      else if (event.type === "live_update") { model.predictions = predict(); model.updateCount += 1; }
    }
    draw();
  }

  function undo() {
    let index = model.events.length - 1;
    while (index >= 0 && !["paint", "erase"].includes(model.events[index].type)) index -= 1;
    if (index < 0) return;
    model.events.splice(index, 1);
    model.events.forEach((event, eventIndex) => { event.seq = eventIndex + 1; });
    recomputeFromEvents();
    status("READY", "idle");
  }

  async function updateModel() {
    if (!model.samples.length) { status("FAIL", "error"); return; }
    const prediction = predict();
    if (!prediction) { status("FAIL", "error"); return; }
    addEvent("live_update", {input_source: "live_update_button"});
    model.predictions = prediction;
    model.updateCount += 1;
    draw();
    status("UPDATED", "active");
  }

  async function postPayload(completed, includeCertify) {
    if (model.submitting) return;
    model.submitting = true;
    if (includeCertify) addEvent("certify", {input_source: "certify_button"});
    const payload = {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      control_condition: model.state.control_condition,
      input_surface: interaction(),
      events: model.events,
      completed,
      world_fingerprint: model.state.plate.world_fingerprint,
    };
    try {
      const response = await fetch("/result", {method: "POST", headers: {"content-type": "application/json"}, body: JSON.stringify(payload)});
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.verdict = "passed";
        status("PASS", "passed");
      } else if (outcome.passed === false) {
        model.verdict = "error";
        if (outcome.state) {
          await render(outcome.state, model.helpers);
          status("FAIL", "error");
        } else status("FAIL", "error");
      }
    } catch (_error) {
      status("FAIL", "error");
    } finally {
      model.submitting = false;
    }
  }

  function resetLocal() {
    model.events = [];
    model.samples = [];
    model.predictions = null;
    model.selectedClass = null;
    model.updateCount = 0;
    document.querySelectorAll("[data-stencil-class]").forEach((button) => button.classList.remove("is-selected"));
    draw();
    status("READY", "idle");
  }

  async function render(state, helpers) {
    document.body.dataset.mechanic = "teach-stencil";
    document.body.dataset.cheatMode = helpers.isCheatMode() ? "true" : "false";
    model.state = state;
    model.helpers = helpers;
    model.canvas = null;
    model.ctx = null;
    model.selectedClass = null;
    model.tool = "paint";
    model.brushSize = Number(state.tools?.default_brush_size || 2);
    model.samples = [];
    model.predictions = null;
    model.events = [];
    model.pointer = null;
    model.drawing = false;
    model.updateCount = 0;
    model.verdict = "idle";
    const mode = interaction();
    helpers.app.innerHTML = `
      <section class="teach-stencil" data-interaction="${helpers.text(mode)}">
        <header class="teach-stencil-header">
          <div><p class="teach-kicker">BOTANICAL PLATE</p><h1>Teach the Stencil</h1><p class="teach-prompt">${helpers.text(state.prompt)}</p></div>
          <div class="teach-contract"><span class="teach-contract-dot"></span><span>${helpers.text(mode.toUpperCase())} INPUT</span></div>
        </header>
        <div class="teach-stencil-grid">
          <aside class="teach-sidebar">
            <div class="teach-panel"><div class="teach-panel-label">MATERIAL PALETTE</div><div class="teach-palette">${state.classes.map((item) => `<button type="button" class="teach-class" data-stencil-class="${Number(item.id)}"><i style="--swatch:${helpers.text(item.swatch)}"></i><span>${helpers.text(item.name)}</span><small>${helpers.text(item.description)}</small></button>`).join("")}</div></div>
            <div class="teach-panel teach-tools"><div class="teach-panel-label">MARKING TOOL</div><div class="teach-tool-row"><button type="button" class="teach-tool is-selected" data-stencil-tool="paint">BRUSH</button><button type="button" class="teach-tool" data-stencil-tool="erase">ERASER</button></div><label class="teach-range">BRUSH SIZE <input id="teach-stencil-size" type="range" min="1" max="5" value="${model.brushSize}"><output id="teach-stencil-size-value">${model.brushSize}</output></label></div>
            <button type="button" class="teach-update" id="teach-stencil-update"><span>↻</span> LIVE UPDATE</button>
            <button type="button" class="teach-certify" id="teach-stencil-certify">CERTIFY MASK</button>
            <button type="button" class="teach-quiet-button" id="teach-stencil-undo">UNDO LAST MARK</button>
            <button type="button" class="teach-quiet-button" id="teach-stencil-fresh">NEW PLATE</button>
          </aside>
          <main class="teach-stage"><div class="teach-stage-top"><span>PLATE VIEW</span></div><canvas id="teach-stencil-canvas" width="900" height="480" aria-label="Botanical plate annotation canvas"></canvas><div class="teach-stage-caption"><span>PLATE</span></div><div class="teach-stencil-verdict readout" data-status="idle">READY</div></main>
        </div>
      </section>`;
    model.canvas = document.querySelector("#teach-stencil-canvas");
    model.ctx = model.canvas.getContext("2d");
    installPointerHandlers();
    document.querySelectorAll("[data-stencil-class]").forEach((button) => button.addEventListener("click", () => chooseClass(button.dataset.stencilClass)));
    document.querySelectorAll("[data-stencil-tool]").forEach((button) => button.addEventListener("click", () => setTool(button.dataset.stencilTool)));
    document.querySelector("#teach-stencil-size")?.addEventListener("input", (event) => { model.brushSize = Number(event.target.value); document.querySelector("#teach-stencil-size-value").textContent = String(model.brushSize); draw(); });
    document.querySelector("#teach-stencil-update")?.addEventListener("click", updateModel);
    document.querySelector("#teach-stencil-certify")?.addEventListener("click", () => postPayload(true, true));
    document.querySelector("#teach-stencil-undo")?.addEventListener("click", undo);
    document.querySelector("#teach-stencil-fresh")?.addEventListener("click", () => postPayload(false, false));
    draw();
  }

  window.teachTheStencilModel = model;
  window.WeirdCaptchaMechanics.teach_the_stencil = {rootSelector: ".teach-stencil", render};
})();

File: weird_captcha_gym/shared_runtime/server/incubator_graders/teach_the_stencil.py
from __future__ import annotations

import math
from typing import Any


MECHANIC_ID = "teach_the_stencil"
_MIN_FULL_PATH_LENGTH = 4.0

_BRUSH_OFFSETS: dict[int, tuple[tuple[int, int], ...]] = {
    1: ((0, 0),),
    2: ((0, 0), (1, 0)),
    3: ((0, 0), (1, 0), (-1, 0), (0, 1), (0, -1)),
    4: ((0, 0), (1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (-1, -1)),
    5: (
        (0, 0), (1, 0), (-1, 0), (0, 1), (0, -1),
        (1, 1), (-1, -1), (1, -1), (-1, 1), (2, 0), (-2, 0), (0, 2), (0, -2),
    ),
}


def _fail(message: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "score": 0, "feedback": message}


def _number(value: Any) -> float | None:
    try:
        result = float(value)
    except (TypeError, ValueError):
        return None
    return result if math.isfinite(result) else None


def _condition(value: Any) -> dict[str, Any]:
    return value if isinstance(value, dict) else {}


def _interaction_sources(interaction: str) -> tuple[str, str]:
    if interaction == "simplified":
        return "proxy_stamp", "proxy_erase"
    if interaction == "full":
        return "brush_drag", "eraser_drag"
    return "", ""


def _brush_size(value: Any) -> int | None:
    if value is None:
        return 1
    try:
        size = int(float(value))
    except (TypeError, ValueError):
        return None
    return size if size in _BRUSH_OFFSETS else None


def _brush_points(point: tuple[float, float], size: int, width: int, height: int) -> list[tuple[int, int]]:
    base_x, base_y = int(math.floor(point[0])), int(math.floor(point[1]))
    result: list[tuple[int, int]] = []
    for offset_x, offset_y in _BRUSH_OFFSETS[size]:
        x, y = base_x + offset_x, base_y + offset_y
        if not (0 <= x < width and 0 <= y < height):
            continue
        if (x, y) not in result:
            result.append((x, y))
    return result


def _eraser_radius(size: int) -> float:
    return 0.8 + 0.55 * float(size)


def _path_length(points: list[tuple[float, float]]) -> float:
    return sum(math.hypot(right[0] - left[0], right[1] - left[1]) for left, right in zip(points, points[1:]))


def _raw_point_key(point: tuple[float, float]) -> tuple[int, int]:
    return int(math.floor(point[0])), int(math.floor(point[1]))


def _feature_distance(feature: list[float], centroid: list[float]) -> float:
    # Color is normalized so the surface texture channel can affect decisions
    # without overwhelming RGB differences.
    return sum((float(feature[i]) / 255.0 - float(centroid[i]) / 255.0) ** 2 for i in range(3)) + 0.65 * (float(feature[3]) - float(centroid[3])) ** 2


def _predict(features: list[list[float]], samples: list[dict[str, Any]], class_count: int) -> list[int]:
    centroids: dict[int, list[float]] = {}
    for class_id in range(class_count):
        points = [item["feature"] for item in samples if int(item["class_id"]) == class_id]
        if points:
            centroids[class_id] = [sum(float(point[channel]) for point in points) / len(points) for channel in range(4)]
    if len(centroids) != class_count:
        return [-1] * len(features)
    return [min(centroids, key=lambda class_id: _feature_distance(feature, centroids[class_id])) for feature in features]


def _valid_point(item: Any, width: int, height: int) -> tuple[float, float] | None:
    if not isinstance(item, dict):
        return None
    x, y = _number(item.get("x")), _number(item.get("y"))
    if x is None or y is None or not (0 <= x < width and 0 <= y < height):
        return None
    return x, y


def _mean_iou(predicted: list[int], target: list[int], class_count: int) -> tuple[float, list[float]]:
    ious: list[float] = []
    for class_id in range(class_count):
        intersection = sum(1 for left, right in zip(predicted, target, strict=True) if left == right == class_id)
        union = sum(1 for left, right in zip(predicted, target, strict=True) if left == class_id or right == class_id)
        ious.append(intersection / union if union else 0.0)
    return (sum(ious) / len(ious) if ious else 0.0), ious


def grade(payload: dict[str, Any], truth: dict[str, Any], public: dict[str, Any]) -> dict[str, Any]:
    if not isinstance(payload, dict) or payload.get("mechanic_id") != MECHANIC_ID:
        return _fail("mechanic mismatch")
    if truth.get("mechanic_id") != MECHANIC_ID or public.get("mechanic_id") != MECHANIC_ID:
        return _fail("task state is not a Teach the Stencil challenge")
    if payload.get("task_id") != truth.get("task_id") or payload.get("challenge_id") != truth.get("challenge_id") or public.get("challenge_id") != truth.get("challenge_id"):
        return _fail("stale task or challenge")
    if payload.get("target_labels") is not None or payload.get("predicted_labels") is not None or payload.get("ground_truth") is not None:
        return _fail("submitted mask data is not an allowed browser action")

    condition = _condition(truth.get("control_condition"))
    if public.get("control_condition") != condition or payload.get("control_condition") != condition:
        return _fail("public interaction condition differs from the stencil contract")
    interaction = str(condition.get("interaction") or "")
    paint_source, erase_source = _interaction_sources(interaction)
    if not paint_source:
        return _fail("invalid stencil interaction condition")

    plate = public.get("plate") or {}
    width, height = int(truth.get("width") or 0), int(truth.get("height") or 0)
    features = plate.get("pixels")
    target = truth.get("target_labels")
    classes = public.get("classes") or []
    class_count = len(classes)
    if width <= 0 or height <= 0 or not isinstance(features, list) or len(features) != width * height or not isinstance(target, list) or len(target) != len(features) or class_count < 2:
        return _fail("malformed stencil state")

    events = payload.get("events")
    if not isinstance(events, list) or not events or len(events) > int(truth.get("sample_budget") or 120) * 8:
        return _fail("training transcript is empty or too long")

    selected: int | None = None
    selected_classes: set[int] = set()
    samples: list[dict[str, Any]] = []
    update_predictions: list[list[int]] = []
    update_changes: list[int] = []
    corrections = 0
    painted_points: set[tuple[int, int]] = set()
    last_action: str | None = None
    update_seen = False
    sample_count = 0

    for expected_sequence, event in enumerate(events, 1):
        if not isinstance(event, dict) or event.get("seq") != expected_sequence:
            return _fail(f"training action {expected_sequence} has an invalid sequence number")
        action = str(event.get("type") or "")
        if action == "select_class":
            try:
                class_id = int(event.get("class_id"))
            except (TypeError, ValueError):
                return _fail("class selection is malformed")
            if class_id not in range(class_count) or event.get("input_source") != "palette_button":
                return _fail("class selection did not use a visible palette button")
            selected = class_id
            selected_classes.add(class_id)
        elif action in {"paint", "erase"}:
            points = event.get("points")
            if not isinstance(points, list) or not points or len(points) > 160:
                return _fail("brush action has no bounded visible point list")
            expected_source = paint_source if action == "paint" else erase_source
            if event.get("input_source") != expected_source:
                return _fail("brush action uses the wrong interaction surface")
            if action == "paint":
                try:
                    class_id = int(event.get("class_id"))
                except (TypeError, ValueError):
                    return _fail("paint action has no class")
                if selected != class_id or class_id not in range(class_count):
                    return _fail("paint action was not bound to the selected visible class")
                selected_classes.add(class_id)
                parsed = [_valid_point(point, width, height) for point in points]
                if any(point is None for point in parsed):
                    return _fail("paint point is outside the visible plate")
                if interaction == "full" and (len(parsed) < 2 or _path_length(parsed) < _MIN_FULL_PATH_LENGTH):
                    return _fail("full brush action must contain a real drag")
                size = _brush_size(event.get("brush_size"))
                if size is None:
                    return _fail("brush size is outside the visible control range")
                event_keys = {_raw_point_key(point) for point in parsed}
                new_event_keys = event_keys.difference(painted_points)
                for x, y in parsed:
                    assert x is not None and y is not None
                    for sample_x, sample_y in _brush_points((x, y), size, width, height):
                        pixel_index = sample_y * width + sample_x
                        feature = features[pixel_index]
                        samples.append({"class_id": class_id, "x": sample_x, "y": sample_y, "feature": feature})
                sample_count += sum(len(_brush_points(point, size, width, height)) for point in parsed)
                if update_seen:
                    if new_event_keys:
                        corrections += 1
                painted_points.update(event_keys)
            else:
                parsed = [_valid_point(point, width, height) for point in points]
                if any(point is None for point in parsed):
                    return _fail("erase point is outside the visible plate")
                if interaction == "full" and (len(parsed) < 2 or _path_length(parsed) < _MIN_FULL_PATH_LENGTH):
                    return _fail("full eraser action must contain a real drag")
                size = _brush_size(event.get("brush_size"))
                if size is None:
                    return _fail("brush size is outside the visible control range")
                radius = _eraser_radius(size)
                for x, y in parsed:
                    assert x is not None and y is not None
                    samples = [item for item in samples if math.hypot(float(item["x"]) - x, float(item["y"]) - y) > radius]
        elif action == "live_update":
            if event.get("input_source") != "live_update_button" or not samples:
                return _fail("Live Update was not triggered from a visible trained model")
            predicted = _predict(features, samples, class_count)
            if predicted[0] < 0:
                return _fail("Live Update requires at least one example for every visible class")
            if update_predictions:
                changed = sum(left != right for left, right in zip(update_predictions[-1], predicted, strict=True))
            else:
                changed = len(predicted)
            update_predictions.append(predicted)
            update_changes.append(changed)
            update_seen = True
        elif action == "certify":
            if event.get("input_source") != "certify_button":
                return _fail("certification did not use the visible button")
            if expected_sequence != len(events):
                return _fail("no actions may follow certification")
        else:
            return _fail(f"unknown stencil action {action!r}")
        last_action = action

    if last_action != "certify" or len(events) < 2 or events[-2].get("type") != "live_update":
        return _fail("the final visible operation must be Live Update immediately before certification")
    if payload.get("completed") is not True:
        return _fail("certification was not completed")
    if sample_count > int(truth.get("sample_budget") or 0):
        return _fail("annotation budget exceeded")
    required = int(truth.get("required_samples_per_class") or 1)
    counts = {class_id: sum(1 for item in samples if int(item["class_id"]) == class_id) for class_id in range(class_count)}
    if selected_classes != set(range(class_count)) or any(counts[class_id] < required for class_id in range(class_count)):
        return _fail("every visible material needs representative labeled examples")
    predicted = update_predictions[-1]
    score, ious = _mean_iou(predicted, [int(item) for item in target], class_count)
    threshold = float(truth.get("threshold") or 1.0)
    passed = score >= threshold
    changed_updates = sum(1 for change in update_changes[1:] if change > 0)
    feedback = f"mean class IoU {score:.3f} (required {threshold:.3f}); updates {len(update_predictions)}; corrections {corrections}; changed updates {changed_updates}; class IoU {[round(value, 3) for value in ious]}"
    return {"graded": True, "passed": passed, "score": round(score * 100, 2), "feedback": feedback}
